codelab/01_kpt-validate/tutorial.md

```diff
@@ -0,0 +1,326 @@
+and deploy the applications to the cluster. 
+
+The <walkthrough-editor-select-line filePath="guestbook-cl/skaffold.yaml" startLine="6" endLine="6" startCharacterOffset="12" endCharacterOffset="20">frontend</walkthrough-editor-select-line> source code is stored in 
+<walkthrough-editor-open-file filePath="guestbook-cl/php-redis/guestbook.php">guestbook-cl/php-redis</walkthrough-editor-open-file> 
+and its configurations are stored in <walkthrough-editor-open-file filePath="guestbook-cl/config/frontend/deployment.yaml">guestbook-cl/config</walkthrough-editor-open-file>.
+
+<walkthrough-pin-section-icon></walkthrough-pin-section-icon>
+**Glossary**
+
+> [1] *`Hydrate`* means rendering a *kustomize* directory or a *kpt* package to a flatten 
+> configuration, each of whose resources contains the full set of the object information.
+
+## Running skaffold
+
+Time to complete: **About 1 minutes**
+
+```bash
+skaffold dev
+```
+    
+`skaffold dev` is the essential skaffold command. It builds the application and then deploys 
+the applications to the bundled cluster. 
+
+Once the deployment is complete, you can exit with `Ctrl+C`.
+
+<walkthrough-notification-menu-icon></walkthrough-notification-menu-icon>
+**Tips**
+
+> `skaffold dev` can automatically detect file changes and kick off a re-deploy.
+> So you don’t need to rerun the command if the file changes.
+
+## Validating the config
+
+Time to complete: **About 10 minutes**
+
+Validating the configurations helps both the app development and devOps to be efficient in a 
+fragile environment. 
+
+This step uses a `kubeval` example to show how `kpt` functions[2] 
+can validate the app configuration and makes the validation itself **as a declarative config.**
+
+*   Download the kpt function resource
+    
+    ```bash
+    kpt pkg get https://github.com/GoogleContainerTools/skaffold.git/codelab/kpt-101/resources/validation-kubeval validation-kubeval
+    ```
+    
+*   Now let's update the skaffold.yaml to use the new validator. Replace the following code with the 
+`.deploy` section in the <walkthrough-editor-select-line filePath="guestbook-cl/skaffold.yaml" startLine="8" endLine="10" startCharacterOffset="0" endCharacterOffset="100">skaffold.yaml</walkthrough-editor-select-line>
+
+    See the full [skaffold.yaml](https://github.com/yuwenma/sample-app/blob/kubeval/skaffold.yaml#L12) 
+    ```yaml
+    deploy:
+      kpt:
+        dir: config
+        fn:
+          fnPath: validation-kubeval
+          network: true
+    ```
+
+-  <walkthrough-editor-select-line filePath="guestbook-cl/skaffold.yaml" startLine="12" endLine="12" startCharacterOffset="6" endCharacterOffset="12">.deploy.kpt.fn.fnPath</walkthrough-editor-select-line> refers to the kpt function directory we just downloaded. 
```

Comment:
I wonder if we should explain a bit about how kpt functions work?

Reply:
Good point! I had the same doubt but my concern is that the kpt fn is not the main focus of the codelab and its concepts extends the topics too far (each codelab step is ~5 mins and it excludes hard-core concepts). So I only added the kpt fn in the *extended reading* at the end. 

Reply:
that makes sense, thanks for clarifying